### PR TITLE
feat(service): compatibility with old service system (initctl)

### PR DESCRIPTION
**Pull request by lparet-yseop**
you can contact me by email : lparet@yseop.com

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -193,50 +193,12 @@
...
-    __tomcat_custom_context_xml.changed or
-    __tomcat_standard_users_xml.changed or
-    __tomcat_custom_users_xml.changed or
-    __tomcat_custom_manager_context_xml.changed or
-    __tomcat_install_libs.changed or
-    __tomcat_install_webapps.changed )
+- name: Install Tomcat Service with systemd (New Version)
+  include: service_systemd_tomcat.yml
+  tags:
+    - new_version
```
Comment (ChristopherDavenport) [inline tasks/main.yml]: Tags are not `when` looped so by default both loops will execute and will have to be programatically known by the runner.

Seems like an ideal place to build out a service matrix, or at least a check to see which is better to run of some sort.

File: tasks/service_ctl_tomcat.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+# tasks file for universal-tomcat - start tomcat with startup.sh script
+
+- name: Install Service
+  template:
+    src: tomcat.conf.j2
+    dest: /etc/init/tomcat.conf
+    mode: 0644
+
+- name: Reload Services To Include Tomcat
+  command: initctl reload-configuration
```
Comment (ChristopherDavenport) [inline tasks/service_ctl_tomcat.yml]: Command Module is Non-Idempotent by default. Definitely needs a `when` clause when the service installation changed. Also does this do the same thing as the service module with `state:reloaded`

File: tasks/service_systemd_tomcat.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+- name: Install Service
+  register: __tomcat_service
+  template:
+    src: tomcat.service.j2
+    dest: /etc/systemd/system/tomcat.service
+    mode: 0644
+
+- name: Start Service
+  register: __tomcat_start_service
+  systemd:
```
Comment (ChristopherDavenport) [inline tasks/service_systemd_tomcat.yml:13]: As I stated above, if this runs every time I believe this will fail the role on a system without systemd.

**Review by ChristopherDavenport — CHANGES_REQUESTED** [review]
Overall I really like these changes and would like to incorporate them. 

Roles should be idempotent, so that they have guaranteed consistency, and owners can know the state of their inventory.

This is definitely a great start!

File: tasks/main.yml (hunk 1 above)
Comment (lparet-yseop) [reply tasks/main.yml]: https://github.com/ChristopherDavenport/ansible-role-universal-tomcat/pull/19/commits/fcc6d71d3affe2a3bcc3dbdd1db1a888e565e489

File: tasks/service_ctl_tomcat.yml (hunk 2 above)
Comment (lparet-yseop) [reply tasks/service_ctl_tomcat.yml]: https://github.com/ChristopherDavenport/ansible-role-universal-tomcat/pull/19/commits/904359218ded7f76c98767763d20e0a310fdfae7

File: tasks/service_systemd_tomcat.yml (hunk 3 above)
Comment (lparet-yseop) [reply tasks/service_systemd_tomcat.yml:13]: https://github.com/ChristopherDavenport/ansible-role-universal-tomcat/pull/19/commits/fcc6d71d3affe2a3bcc3dbdd1db1a888e565e489

**Comment by lparet-yseop** [comment]
I readed your return and I try to fix what you proposed
If you have other returns to improive this role I could do changes myself

I'm learning Ansible and it is a really good training for me 
Thanks !

**Comment by lparet-yseop** [comment]
I commit yet to change the way to check the init system used
with stat on /sbin/init I can know if it's initctl (itsn't a symbolic link) if it's systemd (it's a symbolic link and the source contains "systemd")
For the future, We can add check to detect other init system

**Comment by ChristopherDavenport** [comment]
So if you're willing to go all the way with this, I'm currently pushing an updated Travis build that I will issue as a PR against your build. The build includes multiple Init Based Systems that are currently failing with that current role build.

**Comment by lparet-yseop** [comment]
Few tests failed, I don't know why yet 
But do you thing 10 seconds is sufficient for old OS version ? It's just a hypothesis

**Comment by ChristopherDavenport** [comment]
Easiest way is to take the end of the travis script and run it locally and increase the wait time in `test.yml` to see if there are any changes. I will have more time to investigate this evening in more depth. 

I am working on building a CLI for ansiblecheck to make this easier.

**Comment by lparet-yseop** [comment]
Ok keep me informed if you have more informations
Thank you and have a nice day

**Comment by ChristopherDavenport** [comment]
So here is my debugging for issues like this, and if you are getting frustrated let me know I can give more push. 

The basic problem is that the server is not starting as expected on it systems, even after checking and getting a correct message about the service starting. As a result, I would in order to test put a command that captures the tomcat log file such as  `cat /opt/tomcat/logs/catalina.out` and then put stout_lines as a debug. This allows you to see the internal working of tomcat in regards to your role. You would want to use travis, on a different non-merging branch for that work, or to do it with the docker commands locally. This would allow you to know more about what is occuring with tomcat besides the more obvious information that is is not allowing access to the manager app.

**Comment by appcara-kaminglau** [comment]
I have similar requirement to support non-systemd OS. Perhaps a better way would be to use include_vars for different distro but I think as a quick touch up this works for me.

https://github.com/appcara-kaminglau/ansible-role-universal-tomcat/commit/ab403f6a313aa31a532297cd7fa97bf788bf7c25

**Comment by lparet-yseop** [comment]
The problem is that Docker container can't have systemctl. That's why it works only on recent version of Linux and not with old version (Ubuntu 14 and 12, EL6 and OracleLinux 6)
I tested on classic VM and it's ok...
In my last commit, I tried to create a system which check if we are in a docker container but the solution isn't idempotent. I can't create tomcat as a service... Do you have an idea to do that ?

I am working on VM so for me it works, but here the problem is with Docker...

**Review by ChristopherDavenport — APPROVED** [review]
Here's what we'll do, comment out failing builds and add support with an Experimental tag in the README, since support is still good for systemctl. 

We can work on building something that works more conclusively as a separate issue. The work required to make this operate smoothly may be in [AnsibleCheck](https://github.com/AnsibleCheck/ansiblecheck) rather than this repository. However if we have a working solution I am apt to bring it in and improve moving forward.

When CI passes with those changes we will merge.

**Comment by ChristopherDavenport** [comment]
Do you want me to push the requested fixes in for you? I don't want to step on your toes, but I'd like to get this merged.

**Comment by lparet-yseop** [comment]
Yes if you want, I need it too but I don't want to push errors :)
(Sorry for the response time)

**Comment by ChristopherDavenport** [comment]
I'll get on it ASAP.

**Comment by ChristopherDavenport** [comment]
Sorry it took me so long to get back to this.

**Comment by lparet-yseop** [comment]
It works for me, thk !
Normally I will send you an other PR to add other init system compatibility

**Pull request merged**